File: src/main/java/com/github/njuro/jard/user/token/UserTokenRepository.java
package com.github.njuro.jard.user.token;

import com.github.njuro.jard.user.User;
import java.time.OffsetDateTime;
import java.util.Optional;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.stereotype.Repository;

@Repository
public interface UserTokenRepository extends JpaRepository<UserToken, String> {

  Optional<UserToken> findByUserAndType(User user, UserTokenType type);

  Optional<UserToken> findByUserAndValueAndType(User user, String value, UserTokenType type);

  void deleteByExpirationAtBefore(OffsetDateTime timestamp);
}

File: src/main/kotlin/com/github/njuro/jard/rewrite/user/token/UserTokenRepository.kt
package com.github.njuro.jard.rewrite.user.token

import com.github.njuro.jard.rewrite.user.User
import org.springframework.data.jpa.repository.JpaRepository
import org.springframework.stereotype.Repository
import java.time.OffsetDateTime

@Repository
interface UserTokenRepository : JpaRepository<UserToken, String> {
    fun findByUserAndType(user: User, type: UserTokenType): UserToken?

    fun findByValueAndType(value: String, type: UserTokenType): UserToken?

    fun deleteByUserAndType(user: User, type: UserTokenType)

    fun deleteByExpirationAtBefore(timestamp: OffsetDateTime)

    fun deleteByUser(user: User)
}
